allennlp/models/event2mind.py

```diff
@@ -0,0 +1,467 @@
+                        batch_size,
+                        state.embedder,
+                        state.decoder_cell,
+                        state.output_projection_layer
+                )
+                if target_tokens:
+                    self._update_recall(all_top_k_predictions, target_tokens[name], state.recall)
+                output_dict[f"{name}_top_k_predictions"] = all_top_k_predictions
+                output_dict[f"{name}_top_k_log_probabilities"] = log_probabilities
+
+        return output_dict
+
+    def greedy_search(self,
+                      final_encoder_output: torch.LongTensor,
+                      target_tokens: Dict[str, torch.LongTensor],
+                      target_embedder: Embedding,
+                      decoder_cell: GRUCell,
+                      output_projection_layer: Linear) -> torch.FloatTensor:
+        """
+        Greedily produces a sequence using the provided ``decoder_cell``.
+        Returns the cross entropy between this sequence and ``target_tokens``.
+
+        Parameters
+        ----------
+        final_encoder_output : ``torch.LongTensor``, required
+            Vector produced by ``self._encoder``.
+        target_tokens : ``Dict[str, torch.LongTensor]``, required
+           The output of ``TextField.as_array()`` applied on some target ``TextField``.
+        target_embedder : ``Embedding``, required
+           Used to embed the target tokens.
+        decoder_cell: ``GRUCell``, required
+           The recurrent cell used at each time step.
+        output_projection_layer: ``Linear``, required
+           Linear layer mapping to the desired number of classes.
+        """
+        targets = target_tokens["tokens"]
+        target_sequence_length = targets.size()[1]
+        # The last input from the target is either padding or the end symbol.
+        # Either way, we don't have to process it. (To be clear, we do still
+        # output and compare against the end symbol, but there is no need to
+        # take the end symbol as input to the decoder.)
+        num_decoding_steps = target_sequence_length - 1
+
+        decoder_hidden = final_encoder_output
+        step_logits = []
+        for timestep in range(num_decoding_steps):
+            # See https://github.com/allenai/allennlp/issues/1134.
+            input_choices = targets[:, timestep]
+            decoder_input = target_embedder(input_choices)
+            decoder_hidden = decoder_cell(decoder_input, decoder_hidden)
+            # (batch_size, num_classes)
+            output_projections = output_projection_layer(decoder_hidden)
+            # list of (batch_size, 1, num_classes)
+            step_logits.append(output_projections.unsqueeze(1))
+        # (batch_size, num_decoding_steps, num_classes)
+        logits = torch.cat(step_logits, 1)
+        target_mask = get_text_field_mask(target_tokens)
+        return self._get_loss(logits, targets, target_mask)
+
+    def beam_search(self,
```

Comment:
This method is too large to go untested - perhaps one that you can add would be to check that using k=1 is equal to calling `greedy_search`. You could add this to the test for the model.